paddle/fluid/operators/controlflow/recurrent_op_helper.cc

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/operators/controlflow/recurrent_op_helper.h"
+
+#include <algorithm>
+#include <string>
+#include <unordered_set>
+#include <utility>
+
+#include "paddle/fluid/framework/program_desc.h"
+
+namespace paddle {
+namespace operators {
+
+static constexpr char kStepBlock[] = "sub_block";
+static constexpr char kStepScopes[] = "step_scopes";
+static constexpr char kInputs[] = "inputs";
+static constexpr char kInputsGrad[] = "inputs@GRAD";
+static constexpr char kOutputs[] = "outputs";
+static constexpr char kOutputsGrad[] = "outputs@GRAD";
+static constexpr char kParameters[] = "parameters";
+static constexpr char kHasStates[] = "has_states";
+static constexpr char kExStates[] = "ex_states";
+static constexpr char kStates[] = "states";
+static constexpr char kSkipEagerDeletionVars[] = "skip_eager_deletion_vars";
```

Comment:
I just noticed that we don't have recurrent_op.h ... I prefer to put those into .h file, but if we don't have them, then I will do as you commented.

Reply:
Discussed offline, because while_op_helper.h has same const char[] name but different content, I put those strings into newly created recurrent_op.h and use RecurrentOpBase::kHasStates to refer to it ...